### PR TITLE
[NumPy] add op random.laplace

### PR DESCRIPTION
**Pull request by AntiZpvoh**
## Description ##
Add op random.laplace, which could generate random samples with Laplace Distribution.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/random/np_laplace_op.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      return std::vector<ResourceRequest>{
+        ResourceRequest::kRandom, ResourceRequest::kTempSpace};
+  })
+.set_attr<FCompute>("FCompute<cpu>", NumpyLaplaceForward<cpu>)
+.set_attr<nnvm::FGradient>("FGradient", MakeZeroGradNodes)
+.add_argument("input1", "NDArray-or-Symbol", "Source input")
+.add_argument("input2", "NDArray-or-Symbol", "Source input")
+.add_arguments(NumpyLaplaceParam::__FIELDS__());
+
+NNVM_REGISTER_OP(_npi_laplace_n)
```
Comment (xidulu) [inline src/operator/numpy/random/np_laplace_op.cc]: If you register the `sample_n` version of `laplace` in the backend, you should implement the corresponding frontend under the numpy extension namespace, you could refer to `normal_n` and `uniform_n` in https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/numpy_extension/random.py

File: src/operator/numpy/random/np_laplace_op.cu [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+#include "./np_laplace_op.h"
+
+namespace mxnet {
+namespace op {
+
+NNVM_REGISTER_OP(_npi_laplace)
+.set_attr<FCompute>("FCompute<gpu>", NumpyLaplaceForward<gpu>);
+
```
Comment (xidulu) [inline src/operator/numpy/random/np_laplace_op.cu]: Don't forget to register the GPU version of `NNVM_REGISTER_OP(_npi_laplace_n)` .

File: python/mxnet/symbol/numpy/random.py [hunk 3]
```diff
@@ -290,6 +290,54 @@ def choice(a, size=None, replace=True, p=None, ctx=None, out=None):
...
+    input_type = (isinstance(loc, np_symbol), isinstance(scale, np_symbol))
+    if dtype is None:
+        dtype = 'float32'
+    if ctx is None:
+        ctx = current_context()
+    if size == ():
+        size = None
+    if input_type == (True, True):
+        return _npi.laplace(loc, scale, loc=None, scale=None, size=size,
+                           ctx=ctx, dtype=dtype, out=out)
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/random.py]: alignment

File: python/mxnet/symbol/numpy/random.py [hunk 4]
```diff
@@ -290,6 +290,54 @@ def choice(a, size=None, replace=True, p=None, ctx=None, out=None):
...
+    elif input_type == (False, True):
+        return _npi.laplace(scale, loc=loc, scale=None, size=size,
+                           ctx=ctx, dtype=dtype, out=out)
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/random.py]: alignment

File: python/mxnet/symbol/numpy/random.py [hunk 5]
```diff
@@ -290,6 +290,54 @@ def choice(a, size=None, replace=True, p=None, ctx=None, out=None):
...
+    elif input_type == (True, False):
+        return _npi.laplace(loc, loc=None, scale=scale, size=size,
+                           ctx=ctx, dtype=dtype, out=out)
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/random.py]: alignment

File: python/mxnet/symbol/numpy/random.py [hunk 6]
```diff
@@ -290,6 +290,54 @@ def choice(a, size=None, replace=True, p=None, ctx=None, out=None):
...
+    else:
+        return _npi.laplace(loc=loc, scale=scale, size=size,
+                           ctx=ctx, dtype=dtype, out=out)
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/random.py]: alignment

File: src/operator/numpy/random/np_laplace_op.h [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  dmlc::optional<float> loc;
+  dmlc::optional<float> scale;
+  std::string ctx;
+  int dtype;
+  dmlc::optional<mxnet::Tuple<int>> size;
+  DMLC_DECLARE_PARAMETER(NumpyLaplaceParam) {
+    DMLC_DECLARE_FIELD(loc);
+    DMLC_DECLARE_FIELD(scale);
+    DMLC_DECLARE_FIELD(size)
+        .set_default(dmlc::optional<mxnet::Tuple<int>>())
```
Comment (haojin2) [inline src/operator/numpy/random/np_laplace_op.h]: 2-space indentation.

File: src/api/operator/numpy/random/np_laplace_op.cc [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file np_init_op.cc
+ * \brief Implementation of the API of functions in src/operator/numpy/np_init_op.cc
```
Comment (xidulu) [inline src/api/operator/numpy/random/np_laplace_op.cc]: np_laplace_op.cc

File: src/operator/numpy/random/np_laplace_op.h [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+          s, outputs[0].Size(), param.loc.value(), param.scale.value(),
+          laplace_tensor.dptr_, outputs[0].dptr<OType>());
+    });
+  } else if (inputs.size() == 1U) {
+    // [scalar tensor], [tensor scalar] case
+    int ndim = FillShape(inputs[0].shape_, inputs[0].shape_, outputs[0].shape_,
+                         &new_lshape, &new_lshape, &new_oshape);
+    int scalar_pos;
+    float scalar_value;
+    // int type_flag = param.t;
```
Comment (xidulu) [inline src/operator/numpy/random/np_laplace_op.h]: remove unused code

**Comment by xidulu** [comment]
LGTM regarding the logic.

**Pull request merged**